### PR TITLE
Use `Build::getenv` instead of `env::var*` in anywhere that makes sense

### PR DESCRIPTION
**Pull request by NobodyXu**
So that we would get:
 - recompilation when the env is changed
 - cached env

Also some optimisation

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2736,13 +2736,13 @@ impl Build {
...
             cmd.args.push("-isysroot".into());
-            cmd.args.push(sdk_path.clone());
+            cmd.args.push(OsStr::new(&sdk_path).to_owned());
 
             if let AppleArchSpec::Catalyst(_) = arch {
                 // Mac Catalyst uses the macOS SDK, but to compile against and
                 // link to iOS-specific frameworks, we should have the support
                 // library stubs in the include and library search path.
-                let ios_support = PathBuf::from(sdk_path).join("System/iOSSupport");
+                let ios_support = Path:: new(&sdk_path).join("System/iOSSupport");
```
Comment (NobodyXu) [inline src/lib.rs:2745]:
```suggestion
                let ios_support = Path::new(&sdk_path).join("System/iOSSupport");
```

**Review by madsmtm — APPROVED** [review]
Might make sense to ban `std::env::var` and `std::env::var_os` via. Clippy's [`disallowed_methods`](https://rust-lang.github.io/rust-clippy/master/index.html#/disallowed_methods) functionality? And then explicitly allow it inside the implementation of `getenv`

**Review by thomcc — APPROVED** [review]

**Comment by NobodyXu** [comment]
> Might make sense to ban `std::env::var` and `std::env::var_os` via. Clippy's [`disallowed_methods`](https://rust-lang.github.io/rust-clippy/master/index.html#/disallowed_methods) functionality? And then explicitly allow it inside the implementation of `getenv`

Thanks!

Applied the suggestion, and turns out that the entire `windows_registry` is using `env::var*`

Took me a while to change them to use a generic function.

cc @thomcc @madsmtm

**Comment by NobodyXu** [comment]
cc @madsmtm This PR should cover most env variables in #906

If there's any missing, feel free to open a comment.

**Comment by NobodyXu** [comment]
Updated the code to remove any generics in `windows_regsitry`.

Having generics would make `cc` takes longer to compile and generate larger binary, and I don't think it would give us any improvements to performance.

**Review by thomcc — APPROVED** [review]

File: src/lib.rs [hunk 2]
```diff
@@ -3534,7 +3540,7 @@ impl Build {
         // Loop through PATH entries searching for each toolchain. This ensures that we
         // are more likely to discover the toolchain early on, because chances are good
         // that the desired toolchain is in one of the higher-priority paths.
-        env::var_os("PATH")
```
[the remark this answers is not in the document]
Comment (NobodyXu) [reply src/lib.rs:3537]: Perhaps we can special-case PATH here?

For example, Update the getenv to not emit anything for PATH

[the remark this answers is not in the document]
Comment (tgross35) [reply src/lib.rs:3537]: Just crosslinking, should be fixed in https://github.com/rust-lang/cc-rs/pull/1215

**Pull request merged**